source/common/upstream/upstream_impl.h

```diff
@@ -596,8 +596,26 @@ class BaseDynamicClusterImpl : public ClusterImplBase {
 protected:
   using ClusterImplBase::ClusterImplBase;
 
-  bool updateDynamicHostList(const HostVector& new_hosts, HostVector& current_hosts,
-                             HostVector& hosts_added, HostVector& hosts_removed);
+  /**
+   * Updates the host list of a single priority by reconciling the list of new hosts
+   * with existing hosts.
+   *
+   * @param new_hosts the full lists of hosts in the new configuration.
+   * @param current_priority_hosts the full lists of hosts for the priority to be updated. The list
+   * will be modified to contain the updated list of hosts.
+   * @param hosts_added_to_current_priority will be populated with hosts added to the priority.
+   * @param hosts_removed_from_current_priority will be populated with hosts removed from the
+   * priority.
+   * @param updated_hosts is used to aggregate the new state of all hosts accross priority, and will
+   * be updated with the hosts that remain in this priority after the update.
+   * @return whether the hosts for the priority changed.
+   */
+  bool updateDynamicHostList(const HostVector& new_hosts, HostVector& current_priority_hosts,
+                             HostVector& hosts_added_to_current_priority,
+                             HostVector& hosts_removed_from_current_priority,
+                             std::unordered_map<std::string, HostSharedPtr>& updated_hosts);
+
+  std::unordered_map<std::string, Upstream::HostSharedPtr> all_hosts_;
```

Comment:
Seems like all_hosts_ is only used by updateDynamicHostList. Perhaps it should be private with an explicit method that takes updated_hosts invoked after all the priorities have been updated? My reasoning is that it makes management of all_hosts_ explicitly the responsibility of BaseDynamicClusterImpl.

Reply:
yeah i like that idea